075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,435 @@
+
+The `Kafka` custom resource is extended by adding a new `spec.cruiseControl.autoRebalance` field.
+By default, without such field, no auto-rebalancing runs on scaling.
+
+The `spec.cruiseControl.autoRebalance` field allows to specify a list of rebalancing modes with the corresponding configuration to be used.
+For each entry in the list, the mode is defined by the `spec.cruiseControl.autoRebalance.mode` field (i.e. `add-brokers` or `remove-brokers`) and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource (read later for more details), by using the `spec.cruiseControl.autoRebalance.config` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+
+In the following example, a user can decide to use the same rebalancing configuration when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the same rebalancing configuration when adding or removing brokers
+      - mode: add-brokers
+        config: my-add-remove-brokers-rebalancing-config
+      - mode: remove-brokers
+        config: my-add-remove-brokers-rebalancing-config
+```
+
+Furthermore, it is possible to use the default Cruise Control rebalancing configuration by omitting the corresponding field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the default Cruise Control rebalancing configuration when adding or removing brokers
+      - mode: add-brokers
+      - mode: remove-brokers
+```
+
+The user can even decide to use a different configuration when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using specific configuration when adding brokers
+      - mode: add-brokers
+        config: my-add-brokers-rebalancing-config
+      # using different configuration when removing brokers  
+      - mode: remove-brokers
+        config: my-remove-brokers-rebalancing-config
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
+This way provides the greatest flexibility to describe in which case to run the auto-rebalancing and with which configuration.
+
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.config` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
```

Comment:
Yeah, good point it should and I think we have a couple options here:

* when running an autorebalance, it just logs a warning and uses the default Cruise Control configuration as when the template is not specified at all.
* fails the reconciliation but in this case, when there is a scale up, it ran anyway and the autorebalance should be started in a later reconciliation when the template reference is fixed.

The first one is the simpler one even if it makes a decision for the user who could have just forgot the annotation. The second one could be more complex to handle the state across reconciliations, but looks like better to me. Wdyt?

Reply:
The second option seems like asking for troubles as you would need to carr the scale-up information for possibly a long time until the user fixes it. The first one seems easier. I guess other option is reporting an error and just ignoring the the scaling (as if autorebalance was disabled)

Reply:
> I guess other option is reporting an error and just ignoring the the scaling (as if autorebalance was disabled)

So we assume autorebalance is disabled (because of wrong template reference) but what do you mean by "ignoring the scaling"? I mean, without autorebalance enabled, a scale down could be blocked (because we can't run the rebalancing first) but the scale up would run (without the subsequent rebalancing). Can you clarify your thoughts here please?

Reply:
I mean you would raise the error and do no auto-rebalancing. For scale-up, that will be it and it will be up to the user to do it. For scale-down, it will come up again and again until the user fixes the configuration or until the brokers are empty somehow.

Reply:
Ah ok. I would go this way then. In general, I don't like having the operator making decisions for the user that much (like running rebalancing anyway but with default CC settings). With your suggestion, the user is the owner of their destiny :-D